Batch File/Batch File.sublime-syntax

```diff
@@ -1967,7 +1967,15 @@ contexts:
 ###[ PROTOTYPES ]#############################################################
 
   bol-pop:
-    # Note: empty lines are ignored
+    - match: \n
+      scope: constant.character.escape.linefeed.dosbatch
```

Comment:
A normal `\n` is not an escape character.

As in Bash's line continuation, only the next line after `^` belongs to current line. Arbritary empty lines may not be ignored.

Reply:
> at most a single empty line in between.

What a horrible spec. 🤦

Reply:
> A normal `\n` is not an escape character.

Except in special circumstances, where it's used to capture a linefeed:

```batchfile
::: newline.bat
@setlocal EnableDelayedExpansion & echo off
(set ^"\=^

)
echo Hi.!\!Hello.
```

```console
> newline.bat
Hi.
Hello.
```

Or escape it and use it to construct macros:

```batchfile
::: newline2.bat
@setlocal DisableDelayedExpansion & echo off
(set ^"\n=^^^

)
set ^"hi=(%\n%
    echo Hello%\n%
    echo World!%\n%
)"
%hi%
```

```console
> newline2.bat
Hello
World!
```

> As in Bash's line continuation, only the next line after `^` belongs to current line with at most a single empty line in between. Arbritary empty lines may not be ignored.
> 
> ```
> echo 12
> :: comment^
> 
> echo 13
> echo 14
> ```
> 
> prints ...
> 
> 12 14
> 
> while
> 
> ```
> echo 12
> :: comment^
> 
> 
> echo 13
> echo 14
> ```
> 
> prints...
> 
> 12 13 14

That's not how this mechanism works. When `^` is present and not within quotes, the next token loses its special meaning (if any) and `\n` is no exception (see above).

Reply:
> What a horrible spec. 🤦

Actually it was created without a spec. It is - as child of the 80s - just a mess with lot's of ugly "workarounds" which interpret language constructs into the syntax, which actually don't exist. Comments is one of those as we learn here.

Actually I wonder why someone would still use it for anything else but automation of trivial tasks.

The information about `^` are interesting.

`%= =%` however is an invalid or non-existing variable - just another workaround to mimic comments as `::` are, which are built on the fact cmd ignoring sorts of invalid expressions.

Reply:
> > A normal \n is not an escape character.
>
> Except in special circumstances, where it's used to capture a linefeed: ...

Well, not according to the linked stackoverflow comment.

First `\n` after `^` is stripped. If 2nd character after `^` is an `\n`, too, it is treated literal, which means, it does not terminate the line after initial `^\n`.

That's why 

```cmd
@echo Hello^

World!
```

prints to 

```
Hello
World!
```

instead of interpreting `World!` as command.

That's also the reason why "label like comments" can contain an empty 2nd line.

```cmd
:: comment^

2nd line of comment
echo No comment
```

Due to circumstances described in phase 1, the empty line can even contain spaces or tabs.


However `rem` behaves inconsistent with regards to that.

If a comment consists of only a single token, it works as expected

```
rem token^

2nd line
echo No comment
```

If however it consists of multiple tokens, it does not.

```
rem token1 token2^

echo No comment
```

Well, I am not sure if we really want to handle all the special cases of that syntax, especially as a "single phase" parser such as ST's syntax engine can barely handle all edge cases produced by "multi phase" parsers with sane efforts. That's also true for Markdown.

Reply:
> Actually I wonder why someone would still use it for anything else but automation of trivial tasks.

For the same reason people use POSIX shells - to write simple, ageless scripts that will work anywhere, and can serve as an entry point to more complicated programs.

> The information about `^` are interesting.
> 
> `%= =%` however is an invalid or non-existing variable - just another workaround to mimic comments as `::` are, which are built on the fact cmd ignoring sorts of invalid expressions.

Yes, and it's mainly used in cases like these, where a line must be empty at evaluation time.

> First `\n` after `^` is stripped. If 2nd character after `^` is an `\n`, too, it is treated literal, which means, it does not terminate the line after initial `^\n`.

Indeed, and that's why `\n` is special in this context - because the first one is stripped.

> Due to circumstances described in phase 1, the empty line can even contain spaces or tabs.

That's not true. How did you come to that conclusion?

> Well, I am not sure if we really want to handle all the special cases of that syntax, especially as a "single phase" parser such as ST's syntax engine can barely handle all edge cases produced by "multi phase" parsers with sane efforts. That's also true for Markdown.

I am of the same opinion - handling every case is impossible without something like semantic highlighting. That doesn't mean I can't at least try...